### ISSUE TITLE
how to infer on 1 GPU？

### ISSUE THREAD

**Issue by linquanxu** [matterport/Mask_RCNN#1075]
hello,
When i predicted images with GPU_COUNT = 1 and i found it used CPU, how to use one GPU to infer?

**Comment by pooyad359** [comment]
when you create the model for inference, select gpu as your device
with tf.device("/gpu:0"):
    model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR,config=inference_config)

**Comment by linquanxu** [comment]
thank you for you reply.
I met other question.   
When I predict one image with a non-GPU version (pip3 install tensorflow), the speed is 2.x seconds. However, when I use the GPU version (pip3 install tensorflow_gpu) to predict 2 images, the speed is still 4.x seconds.
(BTW, I observed that the GPUS have been used when the code of GPU version is running).
it seems that GPU is not helpful for predict.
`import os
import sys
import random
import math
import numpy as np
import skimage.io
import tensorflow as tf
import time
#import matplotlib
#import matplotlib.pyplot as plt
#os.environ['CUDA_VISIBLE_DEVICES'] = '6,7'
# Root directory of the project
ROOT_DIR = os.path.abspath("../")

# Import Mask RCNN
sys.path.append(ROOT_DIR)  # To find local version of the library
from mrcnn import utils
import mrcnn.model as modellib
from mrcnn import visualize
# Import COCO config
sys.path.append(os.path.join(ROOT_DIR, "samples/coco/"))  # To find local version
import coco

#%matplotlib inline

# Directory to save logs and trained model
MODEL_DIR = os.path.join(ROOT_DIR, "logs")

# Local path to trained weights file
COCO_MODEL_PATH = os.path.join(ROOT_DIR, "mask_rcnn_coco.h5")
# Download COCO trained weights from Releases if needed
if not os.path.exists(COCO_MODEL_PATH):
    utils.download_trained_weights(COCO_MODEL_PATH)

# Directory of images to run detection on
IMAGE_DIR = os.path.join(ROOT_DIR, "images")

class InferenceConfig(coco.CocoConfig):
    # Set batch size to 1 since we'll be running inference on
    # one image at a time. Batch size = GPU_COUNT * IMAGES_PER_GPU
    GPU_COUNT = 2
    IMAGES_PER_GPU = 1
    #IMAGE_MIN_DIM = 416
    #IMAGE_MAX_DIM = 416

config = InferenceConfig()
config.display()

# Create model object in inference mode.
#model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR, config=config)
with tf.device("/gpu:7"):
    model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR,config=config)
#model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR,config=config)
# Load weights trained on MS-COCO
model.load_weights(COCO_MODEL_PATH, by_name=True)

class_names = ['BG', 'person', 'bicycle', 'car', 'motorcycle', 'airplane',
               'bus', 'train', 'truck', 'boat', 'traffic light',
               'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird',
               'cat', 'dog', 'horse', 'sheep', 'cow', 'elephant', 'bear',
               'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie',
               'suitcase', 'frisbee', 'skis', 'snowboard', 'sports ball',
               'kite', 'baseball bat', 'baseball glove', 'skateboard',
               'surfboard', 'tennis racket', 'bottle', 'wine glass', 'cup',
               'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
               'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza',
               'donut', 'cake', 'chair', 'couch', 'potted plant', 'bed',
               'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote',
               'keyboard', 'cell phone', 'microwave', 'oven', 'toaster',
               'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors',
               'teddy bear', 'hair drier', 'toothbrush']

file_names = next(os.walk(IMAGE_DIR))[2]
image = skimage.io.imread('../images/test.jpg')

# Run detection
t1 = time.time()
results = model.detect([image,image], verbose=1)
print(time.time()-t1)
# Visualize results
r = results[0]
#visualize.display_instances(image, r['rois'], r['masks'], r['class_ids'],class_names, r['scores'])
print(r['rois'])
print(r['scores'])
`

**Comment by cooleel** [comment]
Hello, sorry if it's not relevant to your question, I got some difficulties on trying to run inference mode on CPU, like on a mac pro. Do you mind let me know how did you run prediction on CPU? Did you use the same script you posted here? (probably only change GPU_COUNT=1 ?)
@linquanxu  Thanks.

**Comment by zhouyuangan** [comment]
@linquanxu Have u found a solution, I need help, thanks pls.

**Comment by freezurbern** [comment]
I have found ```CUDA_VISIBLE_DEVICES``` environmental variable to be the best way to select GPU or CPU. I am using the following code in my Jupyter Notebook to select device for training/inferencing. Place this code at the beginning of your Python script. Also note, you will want to install ```tensorflow-gpu```.

# Configure for desired device (pick one)
## GeForce GTX 1660 Ti 6GB
```python
UseAllowGrowth = True
UseGPUnum = "0"
```
## Tesla K20c 4GB
```python
UseAllowGrowth = False
UseGPUnum = "1"
```
## CPU-ONLY (Slow inferencing only)
```python
UseAllowGrowth = True
UseGPUnum = "-1"
```
# Setup before Mask RCNN (required)
```python
import os
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"]= str(UseGPUnum)
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'

if(UseAllowGrowth):
    from tensorflow.compat.v1 import ConfigProto
    from tensorflow.compat.v1 import InteractiveSession
    config = ConfigProto()
    config.gpu_options.allow_growth = True
    session = InteractiveSession(config=config)

# Show GPU info
from tensorflow.python.client import device_lib
device_lib.list_local_devices()
```

**Comment by linquanxu** [comment]
@cooleel @zhouyuangan ，I am sorry I have no GPU for testing this issue now.
